8192647: GClocker induced GCs can starve threads requiring memory leading to OOME

albertnetymk opened the pull request:
Here is an attempt to simplify GCLocker implementation for Serial and Parallel.

GCLocker prevents GC when Java threads are in a critical region (i.e., calling JNI critical APIs). JDK-7129164 introduces an optimization that updates a shared variable (used to track the number of threads in the critical region) only if there is a pending GC request. However, this also means that after reaching a GC safepoint, we may discover that GCLocker is active, preventing a GC cycle from being invoked. The inability to perform GC at a safepoint adds complexity -- for example, a caller must retry allocation if the request fails due to GC being inhibited by GCLocker.

The proposed patch uses a readers-writer lock to ensure that all Java threads exit the critical region before reaching a GC safepoint. This guarantees that once inside the safepoint, we can successfully invoke a GC cycle. The approach takes inspiration from `ZJNICritical`, but some regressions were observed in j2dbench (on Windows) and the micro-benchmark in [JDK-8232575](https://bugs.openjdk.org/browse/JDK-8232575). Therefore, instead of relying on atomic operations on a global variable when entering or leaving the critical region, this PR uses an existing thread-local variable with a store-load barrier for synchronization.

Performance is neutral for all benchmarks tested: DaCapo, SPECjbb2005, SPECjbb2015, SPECjvm2008, j2dbench, and CacheStress.

Test: tier1-8

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8192647](https://bugs.openjdk.org/browse/JDK-8192647): GClocker induced GCs can starve threads requiring memory leading to OOME (**Bug** - P3)


### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Erik Gahlin](https://openjdk.org/census#egahlin) (@egahlin - **Reviewer**) Review applies to [1b6f908b](https://git.openjdk.org/jdk/pull/23367/files/1b6f908b42154f550f95d1490cc537caef62c094)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23367/head:pull/23367` \
`$ git checkout pull/23367`

Update a local copy of the PR: \
`$ git checkout pull/23367` \
`$ git pull https://git.openjdk.org/jdk.git pull/23367/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23367`

View PR using the GUI difftool: \
`$ git pr show -t 23367`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23367.diff">https://git.openjdk.org/jdk/pull/23367.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23367#issuecomment-2624354729)
</details>

albertnetymk commented:
/cc hotspot-gc

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.hpp:
```diff
@@ -27,126 +27,30 @@
-  static volatile jint _jni_lock_count;  // number of jni active instances.
-  static volatile bool _needs_gc;        // heap is filling, we need a GC
-  static uint _total_collections;        // value for _gc_locker collection
+  static Monitor* _lock;
```
Not sure if having this copy/reference to `Heap_lock` makes the code more clear than referencing `Heap_lock` directly. It needs to be `Heap_lock` anyway.

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.hpp:
```diff
@@ -27,126 +27,30 @@
-// The direct lock/unlock calls do not force a collection if an unlock
-// decrements the count to zero. Avoid calling these if at all possible.
+#include "runtime/mutex.hpp"
 
```
Documentation how GCLocker works/is supposed to work is missing here. It's not exactly trivial.

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.cpp:
```diff
@@ -59,138 +59,78 @@ class GCLockerTimingDebugLogger : public StackObj {
-}
+void GCLocker::block() {
+  assert(_lock->is_locked(), "precondition");
+  assert(Atomic::load(&_is_gc_request_pending) == false, "precondition");
```
```suggestion
  assert(!Atomic::load(&_is_gc_request_pending), "precondition");
```

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.cpp:
```diff
@@ -59,138 +59,78 @@ class GCLockerTimingDebugLogger : public StackObj {
-  increment_debug_jni_lock_count();
+void GCLocker::unblock() {
+  assert(_lock->is_locked(), "precondition");
+  assert(Atomic::load(&_is_gc_request_pending) == true, "precondition");
```
```suggestion
  assert(Atomic::load(&_is_gc_request_pending), "precondition");
```

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.inline.hpp:
```diff
@@ -29,29 +29,35 @@
+    OrderAccess::storeload();
+    if (Atomic::load(&_is_gc_request_pending)) {
+      thread->exit_critical();
+      // slow-path
```
```suggestion
```

Not sure what this `slow-path` comment helps with. Maybe it is describing the next method (but it is named very similarly), or this is an attempt to describe the true-block of the if.
In the latter case, it would maybe be better to put this comment at the start of the true-block of the if, and say something more descriptive like `// Another thread is requesting gc, enter slow path.`

Not sure, feel free to ignore, it's just that to me the comment should either be removed or put upwards a line.

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.hpp:
```diff
@@ -27,126 +27,30 @@
-  // This lock count is updated for all operations and is used to
-  // validate the jni_lock_count that is computed during safepoints.
-  static volatile jint _debug_jni_lock_count;
+  static uint64_t _debug_count;
```
Maybe the variable could be named something less generic, indicating what it is counting. Or add a comment.

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.inline.hpp:
```diff
@@ -29,29 +29,35 @@
-    }
-    decrement_debug_jni_lock_count();
+    Atomic::add(&_debug_count, (uint64_t)-1);
+    // Matching the loadload in GCLocker::block
```
```suggestion
    // Matching the loadload in GCLocker::block.
```

tschatzl commented on src/hotspot/share/gc/shared/gcLocker.cpp:
```diff
@@ -59,138 +59,78 @@ class GCLockerTimingDebugLogger : public StackObj {
-  return (cause == GCCause::_gc_locker) &&
-         (_total_collections != total_collections);
+#ifdef ASSERT
+  // Matching the storestore in GCLocker::exit
```
```suggestion
  // Matching the storestore in GCLocker::exit.
```

tschatzl reviewed (changes requested):
* Idk if GCLocker JFR events need to be available in metadata.xml if the VM does not actually ever send it. I think it does not.
Maybe it is used to decode from old recordings, may be worth asking e.g. @egahlin .
* the bot shows a failure that this PR's CR number shows up in the problemlist, that line needs to be deleted as well. Further it would be interesting to see how many retries there are in the allocation loop with these jnilock* stress test.
* another issue, probably todo is that while Parallel GC has the emergency bailout via GC Overhead limit after excessive retries, Serial does not. Which means that it might retry for a long time, which isn't good (while it did earlier if the number of retries due to gclocker exceed that threshold)

egahlin commented:
> * Idk if GCLocker JFR events need to be available in metadata.xml if the VM does not actually ever send it. I think it does not.
>   Maybe it is used to decode from old recordings, may be worth asking e.g. @egahlin .

If the event is not used and the metric is not interesting to have anymore, remove it from metadata.xml, default.jfc, profile.jfc, EventNames.java and delete the TestGCLockerEvent.java file.

albertnetymk reviewed (commented):
> Further it would be interesting to see how many retries there are in the allocation loop with these jnilock* stress test.

I added `QueuedAllocationWarningCount=1` to `test/hotspot/jtreg/vmTestbase/nsk/stress/jni/gclocker/gcl001.java` and saw retry never exceeds 10 for Serial/Parallel.

> Which means that it might retry for a long time

That occurs only when another java thread successfully triggers a gc, advancing the gc-counter, i.e. there is some system-wide progress. Per-thread progress is hard to guarantee, IMO.

dean-long commented:
I like the direction this is taking us, but I think we could go even further and eventually fold the JNI critical region into the existing safepoint mechanism.  To me, the safepoint mechanism already implements a readers-writer lock, with threads states like _thread_in_Java/_thread_in_vm already being "critical regions".   With this change, we have two nested readers-writer locks that a GC needs to acquire.  I think if we made entering and exiting a JNI critical region change the thread state, (probably by introducing a new thread state), then we don't need a separate readers-writer lock for JNI critical region.  However, maybe we don't want to go that far, as the current solution allows us GC-specific implementations and allows each different GC VMOp to decide if it needs to block for JNI critical regions.

egahlin reviewed (approved):
JFR changes look good.

dholmes-ora commented:
> this PR uses an existing thread-local variable with a store-load barrier for synchronization.

@albertnetymk  can you explain how this protocol is intended to work please. I must be missing some higher-level context that provides additional synchronization because use of the per-thread counters is inherently racy.

dholmes-ora commented:
>  I think we could go even further and eventually fold the JNI critical region into the existing safepoint mechanism.

@dean-long you seem to be forgetting why it was folded out in the first place. :) This was performance critical JNI code where the thread-state transitions were too heavyweight and expensive to use. So we keep the thread safepoint-safe (`_thread_in_native`) and have a way to tell the GC to pause whilst we are in these critical regions.

albertnetymk commented:
> can you explain how this protocol is intended to work please. 

When a GC is requested, the `block()` function sets `_is_gc_request_pending` to `true` and then waits until all threads have exited their critical regions. Any thread attempting to enter a critical region during this time will detect the pending GC flag in `enter()` and follow the slow path, effectively waiting until the GC completes. The storeload barrier is critical to ensure that these two variables -- `_is_gc_request_pending` and the thread-local `_jni_active_critical` -- are accessed in the proper order.

> If you think you need an atomic load here, then it would be needed for in_critical() so just add it there.

`in_critical()` is used only by the owning thread, which has exclusive write access. Therefore, its access does not need to be atomic. However, the reads performed by other threads must be atomic, I believe.

dholmes-ora commented:
> in_critical() is used only by the owning thread,

I see code using `thr->in_critical()` which is not obviously being executed by the current thread on itself. But in any case adding the atomic load to `in_critical()` is basically a no-op (loads are atomic) so no need to add a new API just to do that.

dholmes-ora commented:
> The storeload barrier is critical  ...

I'm not sure it is sufficient. I would have expected some full fences to be needed here as this is very similar to the interaction of thread state with safepoints. I will look closer on Monday (sorry).

dholmes-ora commented:
Sorry still on my to-do list.

dholmes-ora reviewed (changes requested):
@albertnetymk I think that to get the correct "dekker duality" in this code you do need to have full fences between the stores and loads, not just a `storeload` barrier.

dean-long commented on src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp:
```diff
@@ -416,7 +382,7 @@ HeapWord* ParallelScavengeHeap::allocate_old_gen_and_record(size_t size) {
 
 HeapWord* ParallelScavengeHeap::mem_allocate_old_gen(size_t size) {
-  if (!should_alloc_in_eden(size) || GCLocker::is_active_and_needs_gc()) {
+  if (!should_alloc_in_eden(size) || GCLocker::is_active()) {
```
I don't understand why we are checking is_active() here.  The value is not reliable if we aren't at a safepoint, and iterating over all threads seems expensive.

albertnetymk replied:
The intention is to avoid blocking java threads if possible, but there is no fundamental reason why it has be to this way. I have removed it for simpler (or less magical) code.

albertnetymk commented:
> But in any case adding the atomic load to in_critical() is basically a no-op (loads are atomic) so no need to add a new API just to do that.

I have removed the new API, and switched to use the original `in_critical()`.

> I think that to get the correct "dekker duality" in this code you do need to have full fences between the stores and loads, not just a storeload barrier.

I have changed to `fence` for simpler reasoning. (In our codebase, the two have the same implementation, so perf should be the same.)

dholmes-ora commented:
> I have removed the new API, and switched to use the original in_critical().

You still need it to be an atomic load together with whatever compiler barriers that implies, otherwise it can be hoisted out of the spin-loop:
```
while (cur->in_critical()) {
  spin_yield.wait();
}
```

dholmes-ora reviewed (commented):
The GCLocker behaviour would be easier to discern if all of the `thread` parameters/variables that have to be the current thread were actually called `current` (with a few suitably placed assertions).

albertnetymk commented:
> You still need it to be an atomic load

Then, maybe the logic is easier to read if the "atomic" access is visible directly from that context, instead of hiding it inside `in_critical`. Therefore, it probably makes more sense to introduce a new API. WDYT?

> The GCLocker behaviour would be easier to discern ...

Renamed to `current_thread` in `enter`,`exit`, and `enter_slow`.

dholmes-ora commented:
> Then, maybe the logic is easier to read if the "atomic" access is visible directly from that context, instead of hiding it inside in_critical. Therefore, it probably makes more sense to introduce a new API. WDYT?

Okay.

albertnetymk commented:
All suggestions/comments are addressed. Tier1-8 pass. It's ready for another round of review.

dholmes-ora reviewed (commented):
Nothing further from me. Thanks

albertnetymk commented:
Thanks for review.

/integrate
